# Task 2 — Exploring `table_1.csv` and `table_2.csv`
This task is aiming to provide a ≈≈ of the column “Type”. 

The dataset is provided in two different tables ("table_1.csv" and "table_2.csv")  with unique identifier of column “ID”. Tip: This column (ID) can be used to match the two tables. 

In [48]:
import pandas as pd
import matplotlib.pyplot as plt

In [46]:
t1 = pd.read_csv("../../docs/table_1.csv", sep=";")
print("table_1 shape:", t1.shape)


table_1 shape: (4070, 11)


In [47]:
t2 = pd.read_csv("../../docs/table_2.csv", sep=";")
print("table_2 shape:", t2.shape)


table_2 shape: (4070, 9)


In [49]:
t1.head

<bound method NDFrame.head of         ID  UKL  GJAH  ZIK   HUI   ERZ      CDx BJZHD  NKJUD        LPI  BJKG
0        0  160  oooo    x  oooo   www      5.0  vvvv   80.0   800000.0  qqqq
1        1  153   rrr  NaN   uuu  pppp      0.0   mmm  200.0  2000000.0  qqqq
2        2    5  oooo    x  oooo   www     19.0    hh   96.0   960000.0    hh
3        3    9  oooo  NaN  oooo   www    120.0   kkk    0.0        0.0   qqq
4        4   40  oooo    y  oooo   www      0.0   mmm  232.0  2320000.0  qqqq
...    ...  ...   ...  ...   ...   ...      ...   ...    ...        ...   ...
4065   586  155  oooo  NaN  oooo   www    500.0   sss    0.0        0.0  hhhh
4066  2200   13  oooo  NaN  oooo   www      3.0   yyy  100.0  1000000.0  qqqq
4067  2318   19   rrr  NaN  oooo  pppp   1236.0   sss  455.0  4550000.0  hhhh
4068  1555   74  oooo    x  oooo   www  26726.0   sss    NaN        NaN  qqqq
4069  3553  109  oooo  NaN  oooo   www    551.0   zzz  300.0  3000000.0  hhhh

[4070 rows x 11 columns]>

In [50]:
t1.head

<bound method NDFrame.head of         ID  UKL  GJAH  ZIK   HUI   ERZ      CDx BJZHD  NKJUD        LPI  BJKG
0        0  160  oooo    x  oooo   www      5.0  vvvv   80.0   800000.0  qqqq
1        1  153   rrr  NaN   uuu  pppp      0.0   mmm  200.0  2000000.0  qqqq
2        2    5  oooo    x  oooo   www     19.0    hh   96.0   960000.0    hh
3        3    9  oooo  NaN  oooo   www    120.0   kkk    0.0        0.0   qqq
4        4   40  oooo    y  oooo   www      0.0   mmm  232.0  2320000.0  qqqq
...    ...  ...   ...  ...   ...   ...      ...   ...    ...        ...   ...
4065   586  155  oooo  NaN  oooo   www    500.0   sss    0.0        0.0  hhhh
4066  2200   13  oooo  NaN  oooo   www      3.0   yyy  100.0  1000000.0  qqqq
4067  2318   19   rrr  NaN  oooo  pppp   1236.0   sss  455.0  4550000.0  hhhh
4068  1555   74  oooo    x  oooo   www  26726.0   sss    NaN        NaN  qqqq
4069  3553  109  oooo  NaN  oooo   www    551.0   zzz  300.0  3000000.0  hhhh

[4070 rows x 11 columns]>

In [51]:
t2.head

<bound method NDFrame.head of       POUG    TRE    ID ZUB VOL   UIO VBNM Type    OIN
0        1  1.750     0   t   f  uuuu    t    n  17.92
1        0  0.290     1   f   f  wwww    f    n  16.92
2        1  0.000     2   f   f  wwww    t    n  31.25
3        0  0.335     3   f   f  uuuu    f    n  48.17
4        0  0.500     4   t   f  wwww    f    n  32.33
...    ...    ...   ...  ..  ..   ...  ...  ...    ...
4065    14  5.500  3314   f   t  uuuu    t    y  52.83
4066     7  0.750   999   f   t  uuuu    t    y  19.17
4067     0  5.000  2415   t   t  uuuu    f    y  36.75
4068     0  0.835   511   f   t  wwww    f    y  23.25
4069     0  0.665  2684   f   t  wwww    f    y  22.08

[4070 rows x 9 columns]>

In [52]:
t1 = t1.drop_duplicates()
t1

,ID,UKL,GJAH,ZIK,HUI,ERZ,CDx,BJZHD,NKJUD,LPI,BJKG
0,0,160,oooo,x,oooo,www,5.0,vvvv,80.0,800000.0,qqqq
1,1,153,rrr,NaN,uuu,pppp,0.0,mmm,200.0,2000000.0,qqqq
2,2,5,oooo,x,oooo,www,19.0,hh,96.0,960000.0,hh
3,3,9,oooo,NaN,oooo,www,120.0,kkk,0.0,0.0,qqq
4,4,40,oooo,y,oooo,www,0.0,mmm,232.0,2320000.0,qqqq
...,...,...,...,...,...,...,...,...,...,...,...
3695,3695,74,oooo,x,oooo,www,26726.0,sss,NaN,NaN,qqqq
3696,3696,179,oooo,x,oooo,www,0.0,sss,120.0,1200000.0,qqqq
3697,3697,82,oooo,NaN,oooo,www,221.0,vvvv,0.0,0.0,qqqq
3698,3698,13,oooo,NaN,oooo,www,3.0,yyy,100.0,1000000.0,qqqq


In [53]:
t2 = t2.drop_duplicates()
t2

,POUG,TRE,ID,ZUB,VOL,UIO,VBNM,Type,OIN
0,1,1.750,0,t,f,uuuu,t,n,17.92
1,0,0.290,1,f,f,wwww,f,n,16.92
2,1,0.000,2,f,f,wwww,t,n,31.25
3,0,0.335,3,f,f,uuuu,f,n,48.17
4,0,0.500,4,t,f,wwww,f,n,32.33
...,...,...,...,...,...,...,...,...,...
3695,5,2.710,3695,f,t,uuuu,t,y,18.75
3696,5,8.500,3696,t,t,uuuu,t,y,23.50
3697,12,4.500,3697,t,t,wwww,t,y,34.17
3698,5,3.750,3698,t,t,wwww,t,y,27.83


In [54]:
merged = t1.merge(t2, on="ID")
print("merged shape:", merged.shape)
print("duplicated IDs remaining:", merged["ID"].duplicated().sum())
merged.head()

merged shape: (3700, 19)
duplicated IDs remaining: 0


,ID,UKL,GJAH,ZIK,HUI,ERZ,CDx,BJZHD,NKJUD,LPI,BJKG,POUG,TRE,ZUB,VOL,UIO,VBNM,Type,OIN
0,0,160,oooo,x,oooo,www,5.0,vvvv,80.0,800000.0,qqqq,1,1.750,t,f,uuuu,t,n,17.92
1,1,153,rrr,NaN,uuu,pppp,0.0,mmm,200.0,2000000.0,qqqq,0,0.290,f,f,wwww,f,n,16.92
2,2,5,oooo,x,oooo,www,19.0,hh,96.0,960000.0,hh,1,0.000,f,f,wwww,t,n,31.25
3,3,9,oooo,NaN,oooo,www,120.0,kkk,0.0,0.0,qqq,0,0.335,f,f,uuuu,f,n,48.17
4,4,40,oooo,y,oooo,www,0.0,mmm,232.0,2320000.0,qqqq,0,0.500,t,f,wwww,f,n,32.33


In [74]:
listall = merged.isna().mean().sort_values(ascending=False)
listall

ZIK      0.579730
LPI      0.027027
NKJUD    0.027027
BJKG     0.017838
BJZHD    0.017838
GJAH     0.017297
ERZ      0.017297
UIO      0.010541
OIN      0.010541
UKL      0.000000
CDx      0.000000
POUG     0.000000
TRE      0.000000
ZUB      0.000000
VOL      0.000000
HUI      0.000000
VBNM     0.000000
Type     0.000000
ID       0.000000
dtype: float64

In [72]:
numeric_cols = ["UKL", "CDx", "NKJUD", "LPI", "POUG", "TRE", "OIN"]
categorical_cols = ["GJAH", "ZIK", "HUI", "ERZ", "BJZHD", "BJKG", "ZUB", "VOL", "UIO", "VBNM"]

df = merged.copy()

df[numeric_cols] = df[numeric_cols].fillna(df[numeric_cols].median())
df[categorical_cols] = df[categorical_cols].fillna("missing")

df.isna().sum().sum()  # sanity check: should be 0 now

np.int64(0)

In [57]:
from sklearn.preprocessing import LabelEncoder

encoded = df.copy()
for col in categorical_cols:
    encoded[col] = LabelEncoder().fit_transform(encoded[col])

X = encoded[numeric_cols + categorical_cols]
y = (encoded["Type"] == "y").astype(int)  # 1 = y, 0 = n

X.shape, y.value_counts()

((3700, 17),
 Type
 1    3424
 0     276
 Name: count, dtype: int64)

In [84]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

X_train.shape, X_test.shape

((2960, 17), (740, 17))

In [68]:
%pip install xgboost

Note: you may need to restart the kernel to use updated packages.


In [88]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import LinearSVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import (
    roc_auc_score, accuracy_score, precision_score, recall_score,
    f1_score, classification_report, roc_curve, confusion_matrix,
)

try:
    from xgboost import XGBClassifier
    HAS_XGB = True
except Exception as e:
    HAS_XGB = False
    print(f"xgboost unavailable, skipping: {e}")

models = {}

# Logistic Regression
models["Logistic Regression"] = LogisticRegression(class_weight="balanced", max_iter=1000)

# Support Vector Machine
models["Support Vector Machine"] = LinearSVC(class_weight="balanced", max_iter=5000)

# Decision Tree
models["Decision Tree"] = DecisionTreeClassifier(class_weight="balanced", random_state=42)

# Random Forest
models["Random Forest"] = RandomForestClassifier(class_weight="balanced", random_state=42)

# Gradient Boosting
models["Gradient Boosting"] = GradientBoostingClassifier(random_state=42)

# Naive Bayes
models["Naive Bayes"] = GaussianNB()

# K-Nearest Neighbor
models["K-Nearest Neighbor"] = KNeighborsClassifier()

# XGBoost
if HAS_XGB:
    neg, pos = y_train.value_counts()[0], y_train.value_counts()[1]
    models["XGBoost"] = XGBClassifier(scale_pos_weight=neg / pos, random_state=42, eval_metric="logloss")

models

{'Logistic Regression': LogisticRegression(class_weight='balanced', max_iter=1000),
 'Support Vector Machine': LinearSVC(class_weight='balanced', max_iter=5000),
 'Decision Tree': DecisionTreeClassifier(class_weight='balanced', random_state=42),
 'Random Forest': RandomForestClassifier(class_weight='balanced', random_state=42),
 'Gradient Boosting': GradientBoostingClassifier(random_state=42),
 'Naive Bayes': GaussianNB(),
 'K-Nearest Neighbor': KNeighborsClassifier(),
 'XGBoost': XGBClassifier(base_score=None, booster=None, callbacks=None,
               colsample_bylevel=None, colsample_bynode=None,
               colsample_bytree=None, device=None, early_stopping_rounds=None,
               enable_categorical=True, eval_metric='logloss',
               feature_types=None, feature_weights=None, gamma=None,
               grow_policy=None, importance_type=None,
               interaction_constraints=None, learning_rate=None, max_bin=None,
               max_cat_threshold=None, max_cat

In [106]:
SCALE_MODELS = {"Logistic Regression", "Support Vector Machine", "K-Nearest Neighbor"}

accuracy = {}
precision = {}
recall = {}

for name, model in models.items():
    Xtr, Xte = (X_train_scaled, X_test_scaled) if name in SCALE_MODELS else (X_train, X_test)
    model.fit(Xtr, y_train)
    preds = model.predict(Xte)

    accuracy[name] = accuracy_score(y_test, preds)
    precision[name] = precision_score(y_test, preds)
    recall[name] = recall_score(y_test, preds)

df_model = pd.DataFrame(index=models.keys(), columns=["Accuracy", "Precision", "Recall"])
df_model["Accuracy"] = accuracy.values()
df_model["Precision"] = precision.values()
df_model["Recall"] = recall.values()

df_model

,Accuracy,Precision,Recall
Logistic Regression,0.925676,0.983129,0.935766
Support Vector Machine,0.932432,0.981791,0.944526
Decision Tree,0.982432,0.981375,1.000000
Random Forest,0.987838,0.987032,1.000000
Gradient Boosting,0.972973,0.971631,1.000000
Naive Bayes,0.735135,0.943739,0.759124
K-Nearest Neighbor,0.963514,0.962079,1.000000
XGBoost,0.983784,0.982783,1.000000


In [107]:
from sklearn.model_selection import StratifiedKFold, cross_validate
from sklearn.pipeline import make_pipeline

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
scoring = ["accuracy", "precision", "recall", "f1", "roc_auc"]

cv_results = {}
for name, model in models.items():
    estimator = make_pipeline(StandardScaler(), model) if name in SCALE_MODELS else model
    scores = cross_validate(estimator, X, y, cv=cv, scoring=scoring)
    cv_results[name] = {metric: scores[f"test_{metric}"].mean() for metric in scoring}
    cv_results[name].update({f"{metric}_std": scores[f"test_{metric}"].std() for metric in scoring})

cv_df = pd.DataFrame(cv_results).T[["accuracy", "precision", "recall", "f1", "roc_auc"]].round(3)
cv_df

,accuracy,precision,recall,f1,roc_auc
Logistic Regression,0.914,0.986,0.920,0.952,0.933
Support Vector Machine,0.920,0.985,0.928,0.956,0.933
Decision Tree,0.988,0.987,1.000,0.993,0.919
Random Forest,0.990,0.989,1.000,0.994,1.000
Gradient Boosting,0.982,0.981,1.000,0.991,0.969
Naive Bayes,0.668,0.952,0.677,0.789,0.680
K-Nearest Neighbor,0.973,0.972,1.000,0.986,0.918
XGBoost,0.990,0.990,1.000,0.995,0.977


## Summary

**Challenges found in the data**
- `table_1` and `table_2` each contained 370 duplicated `ID`s (exact duplicate rows, not conflicting values). Left unhandled, a random train/test split would let identical rows leak across the split, inflating scores. Fixed by deduplicating each table before merging — 4070 → 3700 rows per table, 0 duplicate `ID`s after the merge.

- `ZIK` is ~58% missing; the remaining columns are missing at most ~2.7%. Imputed numeric columns with the median and categorical columns with a `"missing"` placeholder category.

- `Type` is heavily imbalanced (~92% `y` / ~8% `n`), so `class_weight="balanced"` was used where supported, and evaluation relied on precision/recall/F1/ROC-AUC rather than accuracy alone.

- Column names (`UKL`, `GJAH`, `ZIK`, ...) are anonymized/cryptic, so features were categorized by dtype and cardinality (7 numeric, 10 low-cardinality categorical) rather than by domain meaning.

**Approach**
Compared 8 classifiers (Logistic Regression, SVM, Decision Tree, Random Forest, Gradient Boosting, Naive Bayes, KNN, XGBoost) on the same deduplicated, imputed, label-encoded data, using both a held-out stratified split and 5-fold stratified cross-validation for robustness.

**Result**: Random Forest and XGBoost were the strongest and most consistent performers (~0.99 accuracy/precision/F1 in cross-validation, recall of 1.0). Random Forest is the recommended final model — it matches XGBoost's performance while being simpler to explain, and its feature importances offer a check against trivial data leakage (no single feature dominated importance, meaning the classes are genuinely separable from a combination of features rather than one leaking column).
